From 5178d9190012e58923f844c3875908facd720251 Mon Sep 17 00:00:00 2001
From: Hiromitsu Yamasaki <hiromitsu.yamasaki.ym@renesas.com>
Date: Tue, 5 Jun 2018 18:44:40 +0900
Subject: [PATCH 247/909] spi: sh-msiof: Set 2 clock delay for R-Car H3 Ver.3.0
 only

commit a14d0903968f1524304f5b4de3da584778dbdfde from
git://git.kernel.org/pub/scm/linux/kernel/git/horms/renesas-bsp.git

This patch sets 2 clock cycle delay to the DTDL of SITMDR1/SIRMDR1
register. It is a specification of R-Car H3 Ver.3.0 only.

Signed-off-by: Hiromitsu Yamasaki <hiromitsu.yamasaki.ym@renesas.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/spi/spi-sh-msiof.c | 15 ++++++++++++++-
 1 file changed, 14 insertions(+), 1 deletion(-)

diff --git a/drivers/spi/spi-sh-msiof.c b/drivers/spi/spi-sh-msiof.c
index 3feb0b77689e..034d8dcbd45b 100644
--- a/drivers/spi/spi-sh-msiof.c
+++ b/drivers/spi/spi-sh-msiof.c
@@ -2,7 +2,7 @@
  * SuperH MSIOF SPI Master Interface
  *
  * Copyright (c) 2009 Magnus Damm
- * Copyright (C) 2014-2017 Renesas Electronics Corporation
+ * Copyright (C) 2014-2018 Renesas Electronics Corporation
  * Copyright (C) 2014-2017 Glider bvba
  *
  * This program is free software; you can redistribute it and/or modify
@@ -99,6 +99,7 @@ struct sh_msiof_spi_priv {
 #define MDR1_BITLSB_SHIFT	 24 /* MSB/LSB First (1 = LSB first) */
 #define MDR1_DTDL_MASK   0x00700000 /* Data Pin Bit Delay Mask */
 #define MDR1_DTDL_SHIFT		 20 /* Data Pin Bit Delay for MSIOF_SYNC */
+#define MDR1_DTDL_2CLK          200 /*   2-clock-cycle delay */
 #define MDR1_SYNCDL_SHIFT	 16 /* Frame Sync Signal Timing Delay */
 #define MDR1_FLD_MASK	 0x0000000c /* Frame Sync Signal Interval (0-3) */
 #define MDR1_FLD_SHIFT		  2
@@ -203,12 +204,15 @@ struct sh_msiof_spi_priv {
 /* Check LSI revisions and set specific quirk value */
 #define TRANSFER_WORKAROUND_H3WS10  BIT(0) /* H3ES1.0 workaround */
 #define TRANSFER_WORKAROUND_H3WS11  BIT(1) /* H3ES1.1 workaround */
+#define TRANSFER_2CLK_DELAY_H3WS30  BIT(2) /* H3ES3.0 specification */
 
 static const struct soc_device_attribute rcar_quirks_match[]  = {
 	{ .soc_id = "r8a7795", .revision = "ES1.0",
 		.data = (void *)TRANSFER_WORKAROUND_H3WS10, },
 	{ .soc_id = "r8a7795", .revision = "ES1.1",
 		.data = (void *)TRANSFER_WORKAROUND_H3WS11, },
+	{ .soc_id = "r8a7795", .revision = "ES3.0",
+		.data = (void *)TRANSFER_2CLK_DELAY_H3WS30, },
 	{/*sentinel*/},
 };
 
@@ -1508,6 +1512,15 @@ static int sh_msiof_spi_probe(struct platform_device *pdev)
 		}
 	}
 
+	if (p->quirks & TRANSFER_2CLK_DELAY_H3WS30 &&
+	    !spi_controller_is_slave(p->master)) {
+		if (info->dtdl)
+			dev_warn(&pdev->dev,
+				 "Set 2 clock delay for R-Car H3 Ver.3.0 only\n"
+				);
+		info->dtdl = MDR1_DTDL_2CLK;
+	}
+
 	return 0;
 
  err2:
-- 
2.17.1

